### PR TITLE
feat: load configs from yaml file and environment variables

### PR DESCRIPTION
**Pull request by rohilsurana**
Adds a package to load configs from yaml file and env variables.
Removes overhead of defining/mapping keys for both the methods by using struct tags.

### CODE REVIEW

**Comment by rohilsurana** [comment]
There are a few todos in the readme, will create github issues for the same if the PR gets merged.

File: config/config.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+	"github.com/jeremywohl/flatten"
+	"github.com/mcuadros/go-defaults"
+	"github.com/mitchellh/mapstructure"
+	"github.com/spf13/viper"
+)
+
+// LoadConfig loads configuration into the given mapstructure (https://github.com/mitchellh/mapstructure)
+// from a config.yaml file and overrides with any values set in env variables
+func LoadConfig(config interface{}) {
```
Comment (ravisuhag) [inline config/config.go:16]: Let's call this just load. since config is covered in the package.

Comment (rohilsurana) [reply config/config.go]: yeah that makes sense.

Comment (rohilsurana) [reply config/config.go]: made the change

File: config/config.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		if value.Kind() != reflect.Struct {
+			panicf("Require Ptr to a Struct for Load. Got Ptr to %v\n", value.Kind())
+		}
+	}
+}
+
+func getViperWithDefaults() *viper.Viper {
+	v := viper.New()
+	v.SetConfigName("config")
+	v.AddConfigPath("./")
```
Comment (ravisuhag) [inline config/config.go:67]: @rohilsurana We might want to make this configurable.

Comment (rohilsurana) [reply config/config.go:113]: pushed a commit for this

**Comment by ravisuhag** [comment]
@rohilsurana Would it make sense that instead of config package panicking it only returns an error and the package user decides what to do with that error? Are there any specific reasons because of which we are doing panic within package.

File: config/config.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+// Load loads configuration into the given mapstructure (https://github.com/mitchellh/mapstructure)
+// from a config.yaml file and overrides with any values set in env variables
+func (l *Loader) Load(config interface{}) {
+	verifyParamIsPtrToStructElsePanic(config)
+
+	l.v.AutomaticEnv()
+
+	err := l.v.ReadInConfig()
+	if err != nil {
+		fmt.Println(err)
```
Comment (ravisuhag) [inline config/config.go:70]: Even for CLIs this fmt will print config file not found. Either we can allow creating a loader without config file or optional config file. Or maybe we can avoid printing and only return the error. So clients can decide if this maters to them or not. 

e..g 
```
if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
  return c, err
}

Thoughts?

Comment (rohilsurana) [reply config/config.go]: we can avoid printing it

File: config/README.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# config
+
+This package is an wrapper over viper wwith opinionated defaults that allows loading config from a yaml file and environment variables.
```
Comment (kushsharma) [inline config/README.md:3]: `with`

File: config/config.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	}
+}
+
+func WithConfigName(in string) LoaderOption {
+	return func(l *Loader) {
+		l.v.SetConfigName(in)
+	}
+}
+
+func AddConfigPath(in string) LoaderOption {
```
Comment (kushsharma) [inline config/config.go:32]: Should we use same prefixes? `With`

File: config/config.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+func getViperWithDefaults() *viper.Viper {
+	v := viper.New()
+	v.SetConfigName("config")
+	v.AddConfigPath("./")
+	v.SetConfigType("yaml")
+	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+	return v
+}
+
+func getFlattenedStructKeys(config interface{}) (error, []string) {
```
Comment (kushsharma) [inline config/config.go:118]: error should be second argument to make this idiomatic go

File: config/config.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	if err != nil {
+		panicf("unable to get all config keys from struct: %v\n", err)
+	}
+
+	// Bind each conf fields to environment vars
+	for key := range configKeys {
+		err := l.v.BindEnv(configKeys[key])
+		if err != nil {
+			panicf("unable to bind env keys: %v\n", err)
+		}
```
Comment (kushsharma) [inline config/config.go:85]: seeing this pattern of having single arg error statements, we can merge them like
```
if err := l.v.BindEnv(configKeys[key]); err != nil {
    panicf("unable to bind env keys: %v\n", err)
}
```
across all the usage

File: config/README.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	s, _ := json.MarshalIndent(c, "", "  ") // spaces: 2 | tabs: 1 😛
+	fmt.Println(string(s))
+}
+```
+
+In the above program a YAML file or environment variables can be used to configure.
+
+```yaml
+port: 9000
+newrelic:
```
Comment (kushsharma) [inline config/README.md:45]: Will it support a config from env that contains underscore in parent and child? For example
```
port: 9000
new_relic:
  enabled: true
  license_key: base64encodedkey
log_level: debug
```
Now to configure `license_key` what should be the ENV variable?

File: config/README.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    enabled: true
+    appname: config-test-yaml
+    license: ____LICENSE_STRING_OF_40_CHARACTERS_____
+log_level: debug
+```
+
+or
+
+```sh
+export PORT=9001
```
Comment (kushsharma) [inline config/README.md:55]: Can I add prefixes to envs? Like `OPTIMUS_PORT`

File: config/config.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+	for _, option := range options {
+		option(loader)
+	}
+	return loader
+}
+
+// Load loads configuration into the given mapstructure (https://github.com/mitchellh/mapstructure)
+// from a config.yaml file and overrides with any values set in env variables
+func (l *Loader) Load(config interface{}) {
```
Comment (kushsharma) [inline config/config.go:63]: Any reason we choose to `panic` instead of returning an `error` across this method?

File: config/config.go (hunk 5 above)
Comment (rohilsurana) [reply config/config.go:32]: This option can be used multiple times, so you can have multiple paths to check for config.

File: config/README.md (hunk 8 above)
Comment (rohilsurana) [reply config/README.md:45]: The env key would be - `new_relic_license_key`.

File: config/README.md (hunk 9 above)
Comment (rohilsurana) [reply config/README.md:55]: I think we can add this as an option.

File: config/config.go (hunk 10 above)
Comment (rohilsurana) [reply config/config.go:63]: The plan was that if the configuration is not loaded the application can not be started, but will leave that for the application to decide. Will return the error from here for application to handle.

File: config/README.md (hunk 8 above)
Comment (rohilsurana) [reply config/README.md]: updated the example in the README for the same

File: config/config.go (hunk 7 above)
Comment (rohilsurana) [reply config/config.go]: fixed

File: config/config.go (hunk 6 above)
Comment (rohilsurana) [reply config/config.go]: fixed

File: config/config.go (hunk 5 above)
Comment (rohilsurana) [reply config/config.go:38]: Added a comment to function to make it more clear.
Or we can go with`WithAdditionalConfigPath`, wdyt?

File: config/README.md (hunk 4 above)
Comment (rohilsurana) [reply config/README.md]: fixed

File: config/config.go (hunk 3 above)
Comment (rohilsurana) [reply config/config.go]: fixed

File: config/README.md (hunk 9 above)
Comment (rohilsurana) [reply config/README.md]: Added an option for this. Updated README to show it in the example.

**Comment by rohilsurana** [comment]
> @rohilsurana Would it make sense that instead of config package panicking it only returns an error and the package user decides what to do with that error? Are there any specific reasons because of which we are doing panic within package.

yeah I agree it makes sense to leave it to the application to handle the errors. Have refactored to return error from the `Load` method.

File: config/config.go (hunk 5 above)
Comment (ravisuhag) [reply config/config.go:38]: I think even using two or more WithConfigPath will also not be confusing.

**Comment by ravisuhag** [comment]
apart from WIthConfigPath discussion, LGTM.

**Comment by kushsharma** [comment]
LGTM. This is nice, will update Optimus to use this as well.

File: config/config.go (hunk 5 above)
Comment (rohilsurana) [reply config/config.go:38]: Alright then, renaming to `WithConfigPath`.

File: config/config.go (hunk 2 above)
Comment (AkarshSatija) [reply config/config.go:135]: Can we update the readme for the precedence of these config files?

Comment (rohilsurana) [reply config/config.go:135]: This got addressed by - https://github.com/odpf/salt/pull/7

**Pull request merged**